dali/python/nvidia/dali/worker.py

```diff
@@ -0,0 +1,233 @@
+    It keeps track of what shared memory chunks have been already sent and if needed, sends
+    file descriptors of memory chunks that parent process hasn't seen yet.
+
+    Parameters
+    ----------
+    `worker_id` : int
+        Id of the worker passed by the parent process. Added to messages sent over the ``res_pipe``
+        to simplify bookeeping in parent process.
+    `sock` : socket
+        Python wrapper around Unix socket, capable of sending file descriptors between processes.
+    `res_pipe`: pipe
+        Pipe used to send parent process a notification (along with essential meta data info) about
+        ready batch in a given shared memory chunk.
+    """
+
+    def __init__(self, worker_id, sock, res_pipe):
+        self.worker_id = worker_id
+        self.handle_sent = set()
+        self.sock = sock
+        self.res_pipe = res_pipe
+
+    def send(self, processed_tasks: _ProcessedTasks):
+        if processed_tasks.is_failed():  # one of the tasks failed
+            completed_tasks = CompletedTasks.failed(self.worker_id, processed_tasks)
+            self.res_pipe.send(completed_tasks)
+            return
+        batch_serialized = write_batch(processed_tasks.mem_chunk, processed_tasks.data_batch)
+        completed_tasks = CompletedTasks.done(self.worker_id, processed_tasks, batch_serialized)
+        self.res_pipe.send(completed_tasks)
+        # send shared memory handle for underlaying shared memory chunk
+        # if it hasn't been sent ever before
+        mem_chunk_id = batch_serialized.mem_chunk_id
+        if mem_chunk_id not in self.handle_sent:
+            self.handle_sent.add(mem_chunk_id)
+            reduction.send_handle(self.sock, processed_tasks.mem_chunk.shm_chunk.handle, os.getppid())
+
+    def shutdown(self):
+        """Force to close all communication channels (sockets and pipes) to unlock main process
+        in case of error, when it may be waiting for messages that we can't deliver or the
+        state of protocol is mismatched"""
+        self.res_pipe.close()
+        self.sock.shutdown(socket.SHUT_RDWR)
+        self.sock.close()
+
+class CallbackContext:
+    """Worker can run multiple Python callbacks, CallbackContext is used to
+    (independently from other callbacks) manage shared memory used to pass
+    results of the callback calls.
+    """
+
+    def __init__(self, callback, mem_chunks):
+        self.callback = callback
+        self.mem_chunks = mem_chunks
+        self.batch_i = 0
+
+    def _next_batch_i(self):
+        res = self.batch_i
+        self.batch_i = (self.batch_i + 1) % len(self.mem_chunks)
+        return res
+
+    def next_mem_chunk(self):
+        batch_i = self._next_batch_i()
+        return self.mem_chunks[batch_i]
+
+    def close(self):
+        for chunk in self.mem_chunks:
+            chunk.close()
+
+def dispatcher(batch_dispatcher, ready_cv, ready_queue):
```

Comment:
Why isn't it a member of the dispatcher object?

Reply:
Probably to make it easily usable as the worker's thread.

Reply:
done. not sure if that's better.